tpu-provisioner/internal/controller/nodepool_garbage_collector.go

```diff
@@ -0,0 +1,98 @@
+package controller
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/GoogleCloudPlatform/ai-on-gke/tpu-provisioner/internal/cloud"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// NodePoolGarbageCollector deletes node pools that have no Nodes,
+// are in an errored state, and where the Pod that created the node pool
+// no longer exists (the deletion reconciler would not see these b/c there
+// are no Node objects).
+type NodePoolGarbageCollector struct {
+	Interval time.Duration
+	client.Client
+	Provider cloud.Provider
+}
+
+func (g *NodePoolGarbageCollector) Run(ctx context.Context) {
+	log := ctrllog.Log.WithName("nodepool-garbage-collector")
+
+	t := time.NewTicker(g.Interval)
+
+	for {
+		select {
+		case <-ctx.Done():
+			t.Stop()
+			return
+		case <-t.C:
+		}
+
+		log.Info("starting node pool garbage collection loop")
+
+		nodepools, err := g.Provider.ListNodePools()
+		if err != nil {
+			log.Error(err, "failed to list errored node pools")
+			continue
+		}
+
+		for _, np := range nodepools {
+			log := log.WithValues(
+				"nodepool", np.Name,
+				"createdForPodName", np.CreatedForPod.Name,
+				"createdForPodNamespace", np.CreatedForPod.Namespace,
+			)
+
+			if !np.Error {
+				continue
+			}
+
+			if np.CreatedForPod.Name == "" || np.CreatedForPod.Namespace == "" {
+				log.Info("skipping garbage collection of node pool, no pod reference")
+				continue
+			}
+
+			// Check if the Pod that triggered the Node Pool creation still exists.
+			err := g.Get(ctx, np.CreatedForPod, &v1.Pod{})
+			if err == nil {
+				log.Info("skipping garbage collection of node pool, pod still exists",
```

Comment:
What about the case where the pod which triggered node pool creation is still pending, waiting for the nodes to be ready but since the provisioning failed the nodes either don't exist or are in an ERROR state?

I encountered a case today where the pod was still pending on a NP which failed to provision, and the node pool was never deleted so the pod stayed pending and workload never made any progress.

Reply:
Since we are introducing this polling-based approach to NP deletion/clean up, I think we should consolidate the deletion logic to be completely encapsulated in this garbage collector, rather than split between a polling-based garbage collector and reconciliation-based deletion controller. This would simplify things and make it easier to reason about, and since we control the polling interval we won't be spamming the GKE API.

Reply:
> What about the case where the pod which triggered node pool creation is still pending, waiting for the nodes to be ready but the provisioning failed due to resource exhaustion so there are 0 nodes

The garbage collector would not delete the node pool in this case, and I would expect it goes into a repair state.

Reply:
> Since we are introducing this polling-based approach to NP deletion/clean up, I think we should consolidate the deletion logic to be completely encapsulated in this garbage collector, rather than split between a polling-based garbage collector and reconciliation-based deletion controller.

I think there might be some merit to this - but would want to explore in a later PR